src/desul/array.hpp

```diff
@@ -0,0 +1,310 @@
+         }
+      }
+
+      return true;
+   }
+
+   template <std::size_t I, class T, std::size_t N>
+   DESUL_HOST_DEVICE constexpr T& get(array<T, N>& a) noexcept {
+      return a[I];
+   }
+
+   template <std::size_t I, class T, std::size_t N>
+   DESUL_HOST_DEVICE constexpr T&& get(array<T, N>&& a) noexcept {
+      return move(a[I]);
+   }
+
+   template <std::size_t I, class T, std::size_t N>
+   DESUL_HOST_DEVICE constexpr const T& get(const array<T, N>& a) noexcept {
+      return a[I];
+   }
+
+   template <std::size_t I, class T, std::size_t N>
+   DESUL_HOST_DEVICE constexpr const T&& get(const array<T, N>&& a) noexcept {
+      return move(a[I]);
+   }
+
+   template <class T, std::size_t N>
+   DESUL_HOST_DEVICE constexpr void swap(array<T, N>& lhs,
+                                         array<T, N>& rhs)
+                                            noexcept(noexcept(lhs.swap(rhs))) {
+      lhs.swap(rhs);
+   }
+
+   template <class T, std::size_t N>
+   DESUL_HOST_DEVICE constexpr array<std::remove_cv_t<T>, N> to_array(T (&a)[N]) {
+      array<std::remove_cv_t<T>, N> result;
+
+      for (std::size_t i = 0; i < N; ++i) {
+         result[i] = a[i];
+      }
+
+      return result;
+   }
+
+   template <class T, std::size_t N>
+   DESUL_HOST_DEVICE constexpr array<std::remove_cv_t<T>, N> to_array(T (&&a)[N]) {
+      array<std::remove_cv_t<T>, N> result;
+
+      for (std::size_t i = 0; i < N; ++i) {
+         result[i] = move(a[i]);
+      }
+
+      return result;
+   }
+
+#if 0
+   // TODO: Test if this implementation is faster when make_integer_sequence
+   //       is implemented.
+   namespace detail
+   {
+      template <class T, std::size_t N, std::size_t... I>
+      DESUL_HOST_DEVICE constexpr array<std::remove_cv_t<T>, N>
+         to_array_impl(T (&a)[N], index_sequence<I...>)
+      {
+         return {{a[I]...}};
+      }
+
+      template <class T, std::size_t N, std::size_t... I>
+      DESUL_HOST_DEVICE constexpr array<std::remove_cv_t<T>, N>
+         to_array_impl(T (&&a)[N], index_sequence<I...>)
+      {
+         return {{move(a[I])...}};
+      }
+   }
+    
+   template <class T, std::size_t N>
+   DESUL_HOST_DEVICE constexpr array<std::remove_cv_t<T>, N> to_array(T (&a)[N]) {
+      return detail::to_array_impl(a, make_index_sequence<N>{});
+   }
+
+   template <class T, std::size_t N>
+   DESUL_HOST_DEVICE constexpr array<std::remove_cv_t<T>, N> to_array(T (&&a)[N]) {
+      return detail::to_array_impl(std::move(a), make_index_sequence<N>{});
+   }
+#endif
+
+   template <class T, std::size_t N>
+   struct tuple_size<array<T, N>> :
+      std::integral_constant<std::size_t, N>
+   { };
+
+   template <class T>
+   inline constexpr std::size_t tuple_size_v = tuple_size<T>::value;
+
+   template <std::size_t I, class T>
+   struct tuple_element;
+    
+   template <std::size_t I, class T, std::size_t N>
+   struct tuple_element<I, array<T,N>> {
+      using type = T;
+   };
```

Comment:
This should be a specialization of std::tuple_element

Reply:
Fixed.